actors/migration/nv9/top.go

```diff
@@ -2,77 +2,206 @@ package nv9
+	var migrations = map[cid.Cid]StateMigration{
+		builtin2.AccountActorCodeID:          nilMigrator{builtin3.AccountActorCodeID},
+		builtin2.CronActorCodeID:             nilMigrator{builtin3.CronActorCodeID},
+		builtin2.InitActorCodeID:             nilMigrator{builtin3.InitActorCodeID},
+		builtin2.MultisigActorCodeID:         nilMigrator{builtin3.MultisigActorCodeID},
+		builtin2.PaymentChannelActorCodeID:   nilMigrator{builtin3.PaymentChannelActorCodeID},
+		builtin2.RewardActorCodeID:           nilMigrator{builtin3.RewardActorCodeID},
+		builtin2.StorageMarketActorCodeID:    marketMigrator{},
+		builtin2.StorageMinerActorCodeID:     nilMigrator{builtin3.StorageMinerActorCodeID},
+		builtin2.StoragePowerActorCodeID:     nilMigrator{builtin3.StoragePowerActorCodeID},
+		builtin2.SystemActorCodeID:           nilMigrator{builtin3.SystemActorCodeID},
+		builtin2.VerifiedRegistryActorCodeID: nilMigrator{builtin3.VerifiedRegistryActorCodeID},
+	}
+	// Set of prior vresion code CIDs for actors to defer during iteration, for explicit migration afterwards.
+	var deferredCodeIDs = map[cid.Cid]struct{}{
+		// None
+	}
+	if len(migrations)+len(deferredCodeIDs) != 11 {
+		panic(fmt.Sprintf("incomplete migration specification with %d code CIDs", len(migrations)))
+	}
+
+	// Load input and output state trees
+	adtStore := adt3.WrapStore(ctx, store)
+	actorsIn, err := states2.LoadTree(adtStore, stateRootIn) // FIXME actorsRootIn?
 	if err != nil {
 		return cid.Undef, err
 	}
-
-	// Migrate Market actor
-	marketActorIn, found, err := actorsIn.GetActor(builtin.StorageMarketActorAddr)
+	actorsOut, err := states3.NewTree(adtStore)
 	if err != nil {
 		return cid.Undef, err
 	}
-	if !found {
-		return cid.Undef, xerrors.Errorf("could not find market actor in state")
+
+	// Setup synchronization
+	grp, ctx := errgroup.WithContext(ctx)
+	inputCh := make(chan *migrationInput)
+	resultCh := make(chan *migrationResult)
+
+	// Iterate all actors in old state root to generate migration inputs for each non-deferred actor.
+	grp.Go(func() error {
+		defer close(inputCh)
+		return actorsIn.ForEach(func(addr address.Address, actorIn *states2.Actor) error {
+			if _, ok := deferredCodeIDs[actorIn.Code]; ok {
+				return nil // Deferred for explicit migration later.
+			}
+			nextInput := &migrationInput{
+				Address:        addr,
+				Actor:          *actorIn, // Must take a copy, the pointer is not stable.
+				StateMigration: migrations[actorIn.Code],
+			}
+			select {
+			case inputCh <- nextInput:
+			case <-ctx.Done():
+				return ctx.Err()
+			}
+			return nil
+		})
+	})
+
+	// Worker threads run migrations on inputs.
+	var workerWg sync.WaitGroup
+	for i := 0; i < cfg.MaxWorkers; i++ {
+		workerWg.Add(1)
+		grp.Go(func() error {
+			defer workerWg.Done()
+			for input := range inputCh {
+				result, err := migrateOneActor(ctx, store, input, priorEpoch)
+				if err != nil {
+					return err
+				}
+				if result == nil { // Actor is deferred
```

Comment:
This appears to now be handled earlier in the pipeline by never sending the deferred input to the worker thread so this carry over from the previous solution should now never be reached.  It might be worth throwing an error in this case, but maybe it's better to just remove the if statement and let the result processing panic in the unexpected event that a migration returns a nil result.

Reply:
Removed the check